pkg/storage/iterator.go

```diff
@@ -323,15 +344,18 @@ func buildIterators(ctx context.Context, chks map[model.Fingerprint][][]*chunken
 	return result, nil
 }
 
-func buildHeapIterator(ctx context.Context, chks [][]*chunkenc.LazyChunk, filter logql.Filter, direction logproto.Direction, from, through time.Time) (iter.EntryIterator, error) {
+func buildHeapIterator(ctx context.Context, chks [][]*chunkenc.LazyChunk, validChunks map[*chunkenc.LazyChunk]bool, filter logql.Filter, direction logproto.Direction, from, through time.Time) (iter.EntryIterator, error) {
 	result := make([]iter.EntryIterator, 0, len(chks))
-
+	log, ctx := spanlogger.New(ctx, "LokiStore.fetchLazyChunks")
 	// __name__ is only used for upstream compatibility and is hardcoded within loki. Strip it from the return label set.
 	labels := dropLabels(chks[0][0].Chunk.Metric, labels.MetricName).String()
-
 	for i := range chks {
 		iterators := make([]iter.EntryIterator, 0, len(chks[i]))
 		for j := range chks[i] {
+			if _, ok := validChunks[chks[i][j]]; !ok {
```

Comment:
This is exactly where you should filter invalid chunk. But I don't like a map. I prefer that you check a boolean on the chunk. I don't think you need a log also, as this will already be logged by the memchunk.go

Reply:
👍